### PR TITLE
Upgrade Spring (REPO-1295), Quartz (REPO-3049) and remove Hibernate (REPO-2873)

### PR DESCRIPTION
**Pull request by killerboot**
* The classes connected to ibatis/dialect were moved to alfresco-repository.
* Major refactoring was done to TransactionSupportUtil class. It includes change of txn resource management. The new Spring 5 code cleans transaction synchronization once the txn is committed/rolled back. This means that "aftercompletion" listeners cannot manage txn resources they saved in earlier stages. The listeners can be caches, validors, locks, etc. In order to overcome the new behaviour a new ResourceHolder was added. The principle is very much similar to Spring internal resource holder. The object holds the current txn resources in ThreadLocal variable. If the txn is suspended another object is created and linked to previous resource holder so that once the txn is restored all of the resources are still in place. The resource holder is cleaned once all of the "aftercompletion" txn listeners have finished their work.
All the tests for this class are part of repository project.

### CODE REVIEW

File: src/main/java/org/alfresco/util/transaction/TransactionSupportUtil.java [hunk 1]
```diff
@@ -55,14 +54,17 @@
     /**
      * The order of synchronization set to be 100 less than the Hibernate synchronization order
```
Comment (andrei-rebegea) [inline src/main/java/org/alfresco/util/transaction/TransactionSupportUtil.java]: not sure if the comment is still valid now

File: src/main/java/org/alfresco/util/transaction/TransactionSupportUtil.java [hunk 2]
```diff
@@ -130,117 +112,131 @@ public static boolean isActualTransactionActive()
...
     @SuppressWarnings("unchecked")
-    public static <R extends Object> R getResource(Object key)
+    public static <R> R getResource(Object key)
     {
-        // get the synchronization
-        TransactionSynchronizationImpl txnSynch = getSynchronization();
-        // get the resource
-        Object resource = txnSynch.resources.get(key);
-        // done
+        if (TransactionSynchronizationManager.isSynchronizationActive())
```
Comment (derekhulley) [inline src/main/java/org/alfresco/util/transaction/TransactionSupportUtil.java]: Please add a comment on why this block of code is necessary.

**Review by derekhulley — APPROVED** [review]
Some of the side effect code could do with commenting.
I added an additional commit that was not directly related.

File: src/main/java/org/alfresco/util/transaction/TransactionSupportUtil.java (hunk 2 above)
Comment (killerboot) [reply src/main/java/org/alfresco/util/transaction/TransactionSupportUtil.java]: Good point. My first thought was that this check is not necessary, but I soon realized that it is not true.
The comments were added.

**Pull request merged**